save-preprocessor/src/main/kotlin/org/cqfn/save/preprocessor/utils/GitUtil.kt

```diff
@@ -5,25 +5,161 @@
         UsernamePasswordCredentialsProvider(gitDto.username, gitDto.password)
     } else {
         CredentialsProvider.getDefault()
     }
+
+    if (tmpDir.exists() && !tmpDir.list().isNullOrEmpty()) {
+        pullProject(gitDto, tmpDir, userCredentials, branchOrCommit)?.let {
+            return it
+        }
+        // Pull was unsuccessful, clean directory before cloning
+        deleteDirectory(tmpDir)
+        generateDirectory(tmpDir)
+    }
+    log.info("Starting clone project ${gitDto.url} into the $tmpDir")
     return Git.cloneRepository()
         .setURI(gitDto.url)
         .setCredentialsProvider(userCredentials)
         .setDirectory(tmpDir)
+        .call().also { git ->
+            switchBranch(git, gitDto.url, branchOrCommit)
+        }
+}
+
+/**
+ * @param gitDto
+ * @param tmpDir
+ * @param userCredentials
+ * @param branchOrCommit
+ * @return jGit git entity
+ */
+fun pullProject(
+    gitDto: GitDto,
+    tmpDir: File,
+    userCredentials: CredentialsProvider?,
+    branchOrCommit: String?
+): Git? {
+    log.info("Starting pull project ${gitDto.url} into the $tmpDir")
+    val git = Git.open(tmpDir)
+
+    log.debug("Reset all changes in $tmpDir before pull command")
+    git.reset()
+        .setMode(ResetCommand.ResetType.HARD)
+        .call()
+
+    if (!switchBranch(git, gitDto.url, branchOrCommit)) {
+        return null
+    }
+
+    val branchName = git.repository.branch
+
+    try {
+        git.pull()
+            .setCredentialsProvider(userCredentials)
+            .setRemote(Constants.DEFAULT_REMOTE_NAME)
+            .setRemoteBranchName(branchName)
+            .setFastForward(MergeCommand.FastForwardMode.FF)
+            .call()
+    } catch (ex: RefNotAdvertisedException) {
+        log.error("Provided branch $branchName seems to be an detached commit, pull command won't be performed!")
+        return null
+    } catch (ex: GitAPIException) {
+        log.error("Error during pull project: ", ex)
+        return null
+    }
+    log.info("Successfully pull branch $branchName for project ${gitDto.url}")
+    return git
+}
+
+/**
+ * @param git
+ * @param repoUrl
+ * @param branchOrCommit
+ * @return flag, whether the switching was successful
+ */
+@Suppress("FUNCTION_BOOLEAN_PREFIX")
+fun switchBranch(git: Git, repoUrl: String, branchOrCommit: String?): Boolean {
+    val branchName = if (branchOrCommit.isNullOrBlank()) getDefaultBranchName(repoUrl) else branchOrCommit
+    log.info("Start switch branch from ${git.repository.branch} to the $branchName for $repoUrl")
+    branchName ?: run {
+        log.error("Branch name wasn't provided and couldn't get default branch for repo $repoUrl")
+        return false
+    }
+
+    if (git.repository.branch == branchName.replace("${Constants.DEFAULT_REMOTE_NAME}/", "")) {
+        log.warn("Requested branch $branchName for git checkout command equals to the current branch, won't provide any actions")
+        return true
+    }
+
+    try {
+        try {
```

Comment:
Why nested `try`s?

Reply:
If we will get an exception `RefAlreadyExistsException` while calling `checkout(git, branchName, true)`, we need to call this method with flag `false`, which also can throw an exception `RefNotFoundException`